rust/src/websocket/logger.rs

```diff
@@ -0,0 +1,53 @@
+/* Copyright (C) 2023 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::websocket::WebSocketTransaction;
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+use std;
+use suricata_derive::EnumStringU8;
+
+#[derive(EnumStringU8)]
+pub enum WebSocketOpcode {
+    Continuation = 0,
+    Text = 1,
+    Binary = 2,
+    Ping = 8,
+    Pong = 9,
+}
+
+fn log_websocket(tx: &WebSocketTransaction, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("websocket")?;
+    js.set_bool("fin", tx.pdu.fin)?;
+    if let Some(xorkey) = tx.pdu.mask {
+        js.set_uint("mask", xorkey.into())?;
+    }
+    if let Some(val) = web_socket_opcode_string(tx.pdu.opcode) {
```

Comment:
So rather `ToString` and `FromStr` ?
You are my reference for what is idiomatic in rust :-p

Is there a performance cost to do type conversion from integer to enum ?

Reply:
Oh, I misread this.. So it essentially goes from u8 -> WebSocketOpCode -> &str?  My first feeling is the derive macro does seem like a little overkill since `WebSocketOpcode` doesn't seem to be used at all in the code, other than behind the derive macro? But that aside..

I think the derived code should implement a trait, or an impl block rather than a make function.  Normally you might add a `to_str()` method...

```
impl WebSocketOpCode {
    fn to_str(&self) -> 'static &str {
        ...
    }

    // Or as a direct replacement.  Call like WebSocketOpcode::to_str()
    fn to_str() -> Option<&'static str> {
        
    }
}
```

Of course this assumes that you have a constructed `WebSocketOpcode` already, which it probably makes sense that `tx.pdu.opcode` might be an instance of, which could be done with an implementation of

```
impl From<u8> for WebSocketOpcode
```

I guess it feels odd that this enum exists only to generate some bare functions, but is never actually used?

Reply:
>  So it essentially goes from u8 -> WebSocketOpCode -> &str? 

Nope, not using `WebSocketOpCode` itself

> the derive macro does seem like a little overkill

I want to code only once the match between integer value and string...
How do I achieve that without overkill ?
And without risking a typo if I code 2 functions (one stringer and one from str)

> Of course this assumes that you have a constructed WebSocketOpcode already

I do not

>  which it probably makes sense that tx.pdu.opcode might be an instance of

It does not seem to fit for me as`WebSocketOpcode` enum has a limited number of values, and I still want opcode 3 that is unknown to be parsed...
Can/should I improve my enum ? Like adding a case `Unknown(u8)` ?

> I guess it feels odd that this enum exists only to generate some bare functions, but is never actually used?

I agree.


Sum up :
- Can/should I improve my enum ? Like adding a case `Unknown(u8)` ?
- Is there a batter way than derive macro ?

Reply:
> Can/should I improve my enum ? Like adding a case `Unknown(u8)` ?

This makes sense.



> Is there a batter way than derive macro ?

Implement the methods on the enum, like `from_str` and `to_str`?

This looks interesting as well: https://github.com/Peternator7/strum

Reply:
> Implement the methods on the enum, like from_str and to_str?

I will end up with a typo if I code twice the strings...

Why did you not use strum for AppLayerEvent derive ?
Looks quite similar

Reply:
Strum does not seem to work for displaying the value of a case Unknown(u8)

cf code in https://docs.rs/strum_macros/0.25.3/strum_macros/derive.Display.html where `Color::Blue(10)` does only display blue without the 10

Reply:
There is no standard `to_str ` right ?

Should I use https://doc.rust-lang.org/std/string/trait.ToString.html ? That is `to_string(&self) -> String` so that allocates even if it is a static string... I think not

Reply:
> There is no standard `to_str ` right ?
> 
> Should I use https://doc.rust-lang.org/std/string/trait.ToString.html ? That is `to_string(&self) -> String` so that allocates even if it is a static string... I think not

No there isn't, but that doesn't mean you can implement it/derive it directly on the `WebSocketOpCode`.

Reply:
> Strum does not seem to work for displaying the value of a case Unknown(u8)
> 
> cf code in https://docs.rs/strum_macros/0.25.3/strum_macros/derive.Display.html where `Color::Blue(10)` does only display blue without the 10

Ah, I didn't look at it that closely.

Reply:
> Why did you not use strum for AppLayerEvent derive ?
> Looks quite similar

Yeah, they do. `derive(AppLayerEvent)` was added to implement the `AppLayerEvent` trait we use in some generic functions and has some additional Suri only methods like `get_event_info`.

Reply:
You can check next version of the PR ;-)